src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionManager.cs

```diff
@@ -87,25 +87,39 @@ public Task<Connection> TryCreateConnectionAsync(string serviceName, object call
                 return TryGetConnectionFromPoolAsync(serviceName, cancellationToken);
             }
 
-            private async Task<Connection> TryGetConnectionFromPoolAsync(string serviceName, CancellationToken cancellationToken)
+            private async Task<OwnedDisposable<Connection>> TryGetConnectionFromPoolAsync(string serviceName, CancellationToken cancellationToken)
             {
-                var queue = _pools.GetOrAdd(serviceName, _ => new ConcurrentQueue<JsonRpcConnection>());
+                var queue = _pools.GetOrAdd(serviceName, _ => new ConcurrentQueue<OwnedDisposable<Connection>>());
                 if (queue.TryDequeue(out var connection))
                 {
-                    return new PooledConnection(this, serviceName, connection);
+                    try
+                    {
+                        return new OwnedDisposable<Connection>(() => new PooledConnection(this, serviceName, ref connection));
+                    }
+                    finally
+                    {
+                        connection.Dispose();
+                    }
                 }
 
-                var newConnection = (JsonRpcConnection)await TryCreateNewConnectionAsync(serviceName, callbackTarget: null, cancellationToken).ConfigureAwait(false);
-                if (newConnection == null)
+                var newConnection = await TryCreateNewConnectionAsync(serviceName, callbackTarget: null, cancellationToken).ConfigureAwait(false);
+                try
                 {
-                    // we might not get new connection if we are either shutdown explicitly or due to OOP terminated
-                    return null;
-                }
+                    if (newConnection == null)
+                    {
+                        // we might not get new connection if we are either shutdown explicitly or due to OOP terminated
+                        return default;
+                    }
 
-                return new PooledConnection(this, serviceName, newConnection);
+                    return new OwnedDisposable<Connection>(() => new PooledConnection(this, serviceName, ref newConnection));
+                }
+                finally
+                {
+                    newConnection.Dispose();
```

Comment:
💭 @mavasani Commenting out this line doesn't seem to trigger CA2000/CA2013, but I'm not sure why.